Automattic/wp-calypso
# Purchases Page: Refactor existing code for function PurchaseMetaPrice with suggestions

### Attempting a refactor of the PurchaseMetaPrice function due to the convoluted nature with better patterns to include:

#### Proposed Changes

- [x] Use only early returns rather than reassigning period so that it's clear what each block's priority is.
- [x]  Don't assume that the default term is "yearly"; if we cannot figure out the term in any other way, maybe display nothing at all and/or throw an error.
- [x]  Only use bill_period_label if none of the other conditions is found (ie: only if plan.term does not exist, etc.).
- [x]  Add unit tests to cover different products (marketplace, plans with different terms, one time purchases, transfers, included purchases, etc.) and verify their term is what is expected.


Fixes #67133

#### Testing Instructions:



* As it is a Refactor, it should not change anything so ensure that the usual functionality described below on the purchases page just works as expected.
* Goto this page `/purchases/subscriptions/:site/:purchaseID` , if the purchase is billed `per year`/`per month`/`per week`/`per day` then it should under the price column as `$price/period` where period is one of `year/month/week/day`  ![ALoapO.png](https://user-images.githubusercontent.com/552016/203852979-805d6eda-670c-4b6c-bd8a-76046f8bea77.png)
* If the purchase is a domain that came free with the plan then you should see `Free with plan` under the price column ![X3xZPs.png](https://user-images.githubusercontent.com/552016/203853193-e7de615f-ab1c-459e-8233-1aa32712d4b9.png)
* For DIFM products, purchase a DIFM product [from here](https://wordpress.com/do-it-for-me/), now on the purchases page, we should see the one-time fee mentioned followed by the `(one-time)` keyword ![7XRI2m.png](https://user-images.githubusercontent.com/552016/203854100-7b4e8c3a-ae62-475c-b6ee-af54a78f0089.png)
* This PR also includes 7 tests for the various kind of products purchased, so please ensure that there are no failing tests.
